### DISCUSSION TITLE
Nullable date property gets 'object' as runtimeType instead of 'Date'

### DISCUSSION THREAD

**Discussion by kalu06** [mikro-orm/mikro-orm#5286] [Q&A] [answered]
After migrating to v6, I have observed that my nullable date entity properties defined like follows get the runtimeType `'object'`.
```
@Property({ columnType: 'timestamp(3)', nullable: true })
deletedAt: Date | null = null;
```

Here's what I see from the metadata of the Entity:
```
[...]
{
        name: 'deletedAt',
        kind: 'scalar',
        nullable: true,
        getter: false,
        setter: false,
        columnTypes: [Array],
        runtimeType: 'object',
        fieldNames: [Array],
        unsigned: undefined,
        type: 'Date'
},
[...]
```

The runtimeType is correct though when the property is defined like this:
```
@Property({ columnType: 'timestamp(3)', nullable: true })
deletedAt?: Date;
```
```
{
          name: 'deletedAt',
          kind: 'scalar',
          nullable: true,
          getter: false,
          setter: false,
          columnTypes: [Array],
          type: 'Date',
          runtimeType: 'Date',
          fieldNames: [Array],
          unsigned: undefined
},
```

Any idea?
Thanks

**Comment by B4nan** [comment, marked as answer]
reflect-metadata can't infer any union type, it literally returns `object` for `whatever | null` type hint, replace `whatever` with whatever you want, it will work the same, and there is no way around that. 

with that said, its weird that the type is not correctly inferred from your `columnType` option, as v6 is trying to do that. 

what you can always do is to specify the type explicitly:

```ts
@Property({ type: 'Date', columnType: 'timestamp(3)', nullable: true })
deletedAt: Date | null = null;
```

And you might not need the `columnType` at all, this could/should behave the same:

```ts
@Property({ type: 'Date', length: 3, nullable: true })
deletedAt: Date | null = null;
```

(unless you are using postgres, then this would map to `timestamptz(3)` instead of `timestamp(3)` - in that case stick to the explicit `columnType`)

**Reply by B4nan** [reply]
Btw how did you find out about this? property validation? and what vesion are you on?

FYI v5 didn't have the `prop.runtimeType` at all (and the reflect-metadata inference worked exactly the same, since inception)

**Reply by kalu06** [reply]
Indeed, it works with
```
@Property({ type: 'Date', columnType: 'timestamp(3)', nullable: true })
deletedAt: Date | null = null;
```

Thanks @B4nan !

> Btw how did you find out about this? property validation? and what vesion are you on?
While trying to debug the issue, adding console.log in the serializer.

**Reply by B4nan** [reply]
> While trying to debug the issue, adding console.log in the serializer.

You say _the issue_ but that's what I am asking about - you haven't said anything about the problem, just that you see a wrong `runtimeType` (which is rather a symptom/cause, not a problem description) :] Which I believe is currently used only for property validation (plus in entity generator and maybe some JSON query casting in postgres).

**Reply by kalu06** [reply]
You're right, that's not an issue as I described it.

Originally, I opened this discussion because I had a unit test which failed after migrating to v6.
The test failed asserting a serialized datetime property value (typed with `Date | null`):
- In v5, I received an ISO date string like 'YYYY-MM-DDTHH:mm:ssZ'
- In v6, I received a string like 'YYYY-MM-DD HH:mm:ss'

**Reply by B4nan** [reply]
interesting, i will take a closer look on why this doesn't work the same, i think it should. there were quite some changes in how the discovery works on multiple levels, this doesnt seem to be an expected change

**Reply by vladkrasn** [reply]
The issue is still present 
![image](https://github.com/user-attachments/assets/fbb27f19-d3af-43a8-955a-e6191bafea86)

Kinda weird that two neighbouring (timestampz) columns return slightly different date formats because one of them is nullable

**Reply by B4nan** [reply]
Can you provide a complete reproduction?

**Reply by vladkrasn** [reply]
Here you go https://github.com/vladkrasn/mikro-orm-diff-dates

**Reply by B4nan** [reply]
You need to specify the `type` too, `columnType` is not enough. It feels like you ignored what I said in this very answer. reflect-metadata cannot infer the runtime type from `Date | null` in your second property, that's why they are different. Also `timestamptz` is the default mapping for `type: 'Date` so you can skip that.

```ts
@Property()
createdAt!: Date // this can be inferred via reflect-metadata, so no need for explicit `type` option

@Property({ type: 'Date', nullable: true })
expiresAt!: Date | null
```

I guess we could make it work with just the `columnType` too.

**Reply by vladkrasn** [reply]
Oops, sorry!
